proposals/lockout-violation-detection.md

```diff
@@ -1,56 +1,96 @@
-happening.
+`X` is missing from the vote `V`. This is a lockout violation because this
+implies that the validator made a vote where `S` popped off `X`, but the
+lockout on `X` from an earlier vote should have prevented that from happening.
 
 Note for each interval `(S, S + 2^n)` we also need to be able to lookup the
 vote. This is important for being able to pull up the votes later as part of a
 proof.
 
-5) For each validator we also track a rooted set in the database as well to
-catch violating lockouts on rooted slots. We can remove an interval
-`(S, S + 2^n)` from the database once the slot becomes a root add it to a
-rooted set for this validator, and any new votes < root also get added to
+#### Reducing lockout
+Attempting to illegally *reduce* a lockout from an older vote in a newer vote
+is a lockout violation.
+
+The tricky bit of this is determining that one vote `V` is *older* than
+another vote `V'`. To this end, we track a flag `is_vote_earlier` whose usage
+we will highlight in the protocol below.
+
+1. We set `is_vote_earlier = false`.
+2. For each newly seen vote `V` made by a validator `X`, for each slot `S` in
+`V`:
+    - If `S` exists in the database:
+        - Compare the lockout `L_V` in the vote `V` on `S` against the greatest
+        lockout `L_D` in the database for that slot `S` made by that validator `X`.
+            - If `L_V` < `L_D`, set `is_vote_earlier=true`. 
+            - If `L_V` == `L_D`, continue.
+            - If `L_V > L_D`, check if the flag `is_vote_earlier=true`. If so,
+            this implies this vote `V` was *older* than some vote `V'` that
+            comitted the greater `S'` lockout to the database, yet `V` has a
+            lesser lockout on `S`, which means the validator reduced lockout on
+            `S` in a later vote. This is a lockout violation.
+    - If `S` does not exist in the database, the above `Removing lockouts`
+    section describes the protocol that will catch violations.
+
+#### Reducing roots
+Reducing the root from earlier to later votes is a lockout violation. We detect
+this as follows:
+
+1. For each validator we track a rooted set in the database. We can remove an
+interval `(S, S + 2^n)` from the database once the slot becomes a root add it
+to a rooted set for this validator, and any new votes < root also get added to
 rooted set.
 
-When we see a vote with root N on the main fork, then we remove all intervals
-`(M, P)` where `M < N && P >= N` and  add `M` to the rooted set.
+2. When we see a vote with root N on the main fork, then we remove all
+intervals `(M, P)` where `M < N && P >= N` and  add `M` to the rooted set.
 
 So for example if we see:
 (Assume `{slot: confirmation count}` format)
-1. `{root: 0, 1: 4, 3: 3, 5: 2, 7: 1}`
-2. `{root: 5, 7: 1}`
-3. Then we add `{1, 3`} to rooted set and remove their intervals from the
+- `{root: 0, 1: 4, 3: 3, 5: 2, 7: 1}`
+- `{root: 5, 7: 1}`
```

Comment:
This example reminds me, might want to add a note somewhere that artificially increasing your lockout (here we root 5 however 7 still has a conf count of 1) is not a slashable offense (in line with the vote program as well).

Reply:
added a note